File: src/sim_rtu/thermal/weather.py
"""Outdoor temperature profile generators.

Provides synthetic outdoor air temperature profiles for simulation:
- Sine wave: diurnal cycle with configurable mean, amplitude, and phase
- Constant: fixed temperature (useful for testing)
- CSV file: replay recorded weather data

The sine wave model uses:
    T(t) = mean + amplitude * sin(2*pi*(t - phase_offset) / 24)
where t is the hour of day.
"""

from __future__ import annotations

from abc import ABC, abstractmethod


class WeatherProfile(ABC):
    """Abstract base class for outdoor temperature profiles."""

    @abstractmethod
    def temperature_at(self, sim_time_hours: float) -> float:
        """Return outdoor air temperature at the given simulation time.

        Args:
            sim_time_hours: Simulation time in hours from start.

        Returns:
            Outdoor air temperature in degF.
        """


class SineWaveWeather(WeatherProfile):
    """Diurnal sine wave outdoor temperature profile.

    Args:
        mean: Average daily temperature (degF).
        amplitude: Temperature swing amplitude (degF).
        phase_offset: Hour of day for peak temperature.
    """

    def __init__(
        self,
        mean: float = 85.0,
        amplitude: float = 15.0,
        phase_offset: float = 14.0,
    ) -> None:
        self._mean = mean
        self._amplitude = amplitude
        self._phase_offset = phase_offset

    def temperature_at(self, sim_time_hours: float) -> float:
        """Return outdoor temperature at the given simulation time."""
        raise NotImplementedError("TODO: implement sine wave calculation")


class ConstantWeather(WeatherProfile):
    """Fixed outdoor temperature (useful for testing).

    Args:
        temperature: Constant outdoor temperature (degF).
    """

    def __init__(self, temperature: float = 85.0) -> None:
        self._temperature = temperature

    def temperature_at(self, sim_time_hours: float) -> float:
        """Return the fixed outdoor temperature."""
        return self._temperature

File: internal/thermal/weather.go
package thermal

import (
	"fmt"
	"math"

	"github.com/VOLTTRON/sim-rtu/internal/config"
)

// WeatherProfile provides outdoor temperature at a given simulation time.
type WeatherProfile interface {
	TemperatureAt(simTimeHours float64) float64
}

// ConstantWeather returns a fixed temperature at all times.
type ConstantWeather struct {
	Temperature float64
}

// TemperatureAt returns the constant temperature.
func (c ConstantWeather) TemperatureAt(_ float64) float64 {
	return c.Temperature
}

// SineWaveWeather generates a sinusoidal temperature pattern over 24 hours.
type SineWaveWeather struct {
	Mean        float64
	Amplitude   float64
	PhaseOffset float64 // hours
}

// TemperatureAt returns the temperature at the given simulation time.
// Formula: Mean + Amplitude * sin(2*pi*(simTimeHours - PhaseOffset) / 24.0)
func (s SineWaveWeather) TemperatureAt(simTimeHours float64) float64 {
	return s.Mean + s.Amplitude*math.Sin(2*math.Pi*(simTimeHours-s.PhaseOffset)/24.0)
}

// NewWeatherProfile creates a WeatherProfile from configuration.
func NewWeatherProfile(cfg *config.WeatherConfig) (WeatherProfile, error) {
	if cfg == nil {
		return ConstantWeather{Temperature: 72.0}, nil
	}

	switch cfg.Type {
	case "static", "constant":
		return ConstantWeather{Temperature: cfg.Temperature}, nil
	case "sine_wave", "sinusoidal":
		return SineWaveWeather{
			Mean:        cfg.Mean,
			Amplitude:   cfg.Amplitude,
			PhaseOffset: cfg.PhaseOffset,
		}, nil
	default:
		return nil, fmt.Errorf("unknown weather type %q", cfg.Type)
	}
}
